auth/src/AzureDevOpsSubscriptionProvider.ts

```diff
@@ -0,0 +1,233 @@
+ * Reference: https://learn.microsoft.com/en-us/entra/workload-id/workload-identity-federation
+ */
+export class AzureDevOpsSubscriptionProvider implements AzureSubscriptionProvider {
+    private _tokenCredential: TokenCredential & { tenantId?: string } | undefined;
+    /**
+    * The resource ID of the Azure DevOps federated service connection,
+    *   which can be found on the `resourceId` field of the URL at the address bar
+    *   when viewing the service connection in the Azure DevOps portal
+    */
+    private _SERVICE_CONNECTION_ID: string;
+    /**
+    * The `Tenant ID` field of the service connection properties
+    */
+    private _DOMAIN: string;
+    /**
+    * The `Service Principal Id` field of the service connection properties
+    */
+    private _CLIENT_ID: string;
+
+    public constructor() {
+        const SERVICE_CONNECTION_ID = process.env.AzCodeServiceConnectionID;
+        const DOMAIN = process.env.AzCodeServiceConnectionDomain;
+        const CLIENT_ID = process.env.AzCodeServiceConnectionClientID;
+
+        if (!SERVICE_CONNECTION_ID || !DOMAIN || !CLIENT_ID) {
+            throw new Error(`Azure DevOps federated service connection is not configured\n
+            process.env.AzCodeServiceConnectionID: ${SERVICE_CONNECTION_ID ? "✅" : "❌"}\n
+            process.env.AzCodeServiceConnectionDomain: ${DOMAIN ? "✅" : "❌"}\n
+            process.env.AzCodeServiceConnectionClientID: ${CLIENT_ID ? "✅" : "❌"}\n
+        `);
+        }
+
+        this._SERVICE_CONNECTION_ID = SERVICE_CONNECTION_ID;
+        this._DOMAIN = DOMAIN;
+        this._CLIENT_ID = CLIENT_ID;
+    }
+
+    async getSubscriptions(_filter: boolean): Promise<AzureSubscription[]> {
+        // ignore the filter setting because not every consumer of this provider will use the Resources extension
+        const results: AzureSubscription[] = [];
+        for (const tenant of await this.getTenants()) {
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            const tenantId = tenant.tenantId!;
+            results.push(...await this.getSubscriptionsForTenant(tenantId));
+        }
+        const sortSubscriptions = (subscriptions: AzureSubscription[]): AzureSubscription[] =>
+            subscriptions.sort((a, b) => a.name.localeCompare(b.name));
+
+        return sortSubscriptions(results);
+    }
+
+    public async isSignedIn(): Promise<boolean> {
+        return !!this._tokenCredential;
+    }
+
+    public async signIn(): Promise<boolean> {
+        this._tokenCredential = await getTokenCredential(this._SERVICE_CONNECTION_ID, this._DOMAIN, this._CLIENT_ID);
+        return !!this._tokenCredential;
+    }
+
+    public async signOut(): Promise<void> {
+        this._tokenCredential = undefined;
+    }
+
+    public async getTenants(): Promise<TenantIdDescription[]> {
+        return [{
+            tenantId: this._tokenCredential?.tenantId,
```

Comment:
So there can only ever be one tenant associated with the token? I also don't see where the `tenantId` property is ever set.